src/liballoc/collections/btree/map.rs

```diff
@@ -2829,8 +2829,8 @@ enum UnderflowResult<'a, K, V> {
     Stole(bool),
 }
 
-fn handle_underfull_node<K, V>(
-    node: NodeRef<marker::Mut<'_>, K, V, marker::LeafOrInternal>,
+fn handle_underfull_node<'a, K: 'a, V: 'a>(
+    node: NodeRef<marker::Mut<'a>, K, V, marker::LeafOrInternal>,
```

Comment:
What is this change about?

Reply:
You did similar things elsewhere. I assume this is important; could you add a comment explaining why?

Reply:
I did it because the compiler told me to, when I turned raw pointer use into reference use, and then it trickled down throughout the code, next to places where the pattern existed already. I can't really comment because I still don't understand whether it means that keys and values have to live longer than `a, or the other way around, and even what either of those mean. I should understand, but there's a lot I should.

Reply:
> I did it because the compiler told me to, when I turned raw pointer use into reference use, and then it trickled down throughout the code, next to places where the pattern existed already.

Okay, fair.

> I can't really comment because I still don't understand whether it means that keys and values have to live longer than `a, or the other way around, and even what either of those mean. I should understand, but there's a lot I should.

It means that the types `K` and `V` must not contain any borrows that are shorter than `'a`. IOW, keys and values have to live longer than ("outlive") `'a`. But this is indeed a tricky notion to deal with -- what I explained only talks about the syntactic types; it is very hard to turn this into a statement about the values described by those types.

Reply:
OK, that helps. Are the places (still) not listing those bounds being sloppy, because they circumvent the borrow checker but they really assume the same, or just nice and terse?

Reply:
These are certainly extra assumptions, otherwise you would not get errors when removing them.

I suspect they are meant to inform the borrow checker about requirements of the unsafe code, thus using the borrow checker to help avoid bugs here -- similar to the Imm/Mut distinction for handles.